src/main/scala/wasm/MiniWasmFX.scala

```diff
@@ -63,124 +66,132 @@ case class EvaluatorFX(module: ModuleInstance) {
+      case ReturnCall(f) => 
+        // System.err.println(s"[DEBUG] return call: $f")
+        evalCall1(f, stack, frame, kont, trail, mkont, brTable, hs, true)
       case RefFunc(f)    =>
         // TODO: RefFuncV stores an applicable function, instead of a syntactic structure
-        kont(RefFuncV(f) :: stack, trail, mkont)
+        kont(RefFuncV(f) :: stack, trail, mkont, hs)
 
       // resumable try-catch exception handling:
       // NOTE(GW): so far we haven't use trail at all, could consider removing it
       case TryCatch(es1, es2) =>
         val newHandler: Handler[Ans] = (s1, k1, _, m1) => evalList(es2, s1, frame, k1, List(), m1, List(), hs)
-        val m1: MCont[Ans] = (s1) => kont(s1, List(), mkont)
+        val m1: MCont[Ans] = (s1) => kont(s1, List(), mkont, hs)
         evalList(es1, List(), frame, initK[Ans], List(), m1, List(), List((-1, newHandler)) ++ hs)
       case Resume0() =>
         val (resume: ContV[Ans]) :: newStack = stack
         resume.k(List(), kont, List(), mkont, List())
       case Throw() =>
         val err :: newStack = stack
-        def kr(s: Stack, k1: Cont[Ans], t1: List[Cont[Ans]], m1: MCont[Ans], hs: Handlers[Ans]): Ans =
-          kont(s, List(), s1 => k1(s1, List(), m1))
+        def kr(s: Stack, k1: Cont[Ans], t1: Trail[Ans], m1: MCont[Ans], h1: Handlers[Ans]): Ans =
+          kont(s, List(), s1 => k1(s1, List(), m1, hs), hs)
         hs.head._2(List(err, ContV(kr)), initK[Ans], List(), mkont)
 
       // WasmFX effect handlers:
       case ContNew(ty) =>
         val RefFuncV(f) :: newStack = stack
-        def kr(s: Stack, k1: Cont[Ans], t1: List[Cont[Ans]], m1: MCont[Ans], hs: Handlers[Ans]): Ans = {
-          evalCall1(f, s, frame/*?*/, k1, List(), m1, List(), hs, false)
+        def kr(s: Stack, k1: Cont[Ans], t1: Trail[Ans], m1: MCont[Ans], hs: Handlers[Ans]): Ans = {
+          evalCall1(f, s, frame/*?*/, k1, t1, m1, List(), hs, false)
         }
-        kont(ContV(kr) :: newStack, List(), mkont)
+        kont(ContV(kr) :: newStack, trail, mkont, hs)
       case Suspend(tagId) =>
         val FuncType(_, inps, out) = module.tags(tagId)
         val (inputs, restStack) = stack.splitAt(inps.size)
-        val kr = (s: Stack, k1: Cont[Ans], t1: List[Cont[Ans]], m1: MCont[Ans], hs: Handlers[Ans]) => {
-          // FIXME: handlers are lost here
-          kont(s ++ restStack, List(), s1 => k1(s1, List(), m1)) // mkont lost here
+        // System.err.println(s"[DEBUG] handlers: $hs")
+        // System.err.println(s"[DEBUG] trail: $trail")
+        val kr = (s: Stack, _: Cont[Ans], t1: Trail[Ans], m1: MCont[Ans], hs1: Handlers[Ans]) => {
+          val index = trail.indexWhere { case (_, tags) => tags.contains(tagId) }
+          val newTrail = if (index >= 0) trail.take(index) else trail
+          kont(s ++ restStack, newTrail ++ t1, m1, hs1) // mkont lost here, and it's safe if we never modify it
```

Comment:
If we replace `newTrail` to just `trail.dropRight(1)`. `nested-strip.wat` test is the only one that will fail